internal/transport/transport_test.go

```diff
@@ -2136,6 +2137,69 @@ func (s) TestHeadersHTTPStatusGRPCStatus(t *testing.T) {
 	}
 }
 
+func (s) TestWriteHeaderConnectionError(t *testing.T) {
+	server, client, cancel := setUp(t, 0, notifyCall)
+	defer cancel()
+	defer server.stop()
+
+	waitWhileTrue(t, func() (bool, error) {
+		server.mu.Lock()
+		defer server.mu.Unlock()
+
+		if len(server.conns) == 0 {
+			return true, fmt.Errorf("timed-out while waiting for connection to be created on the server")
+		}
+		return false, nil
+	})
+
+	if len(server.conns) != 1 {
+		t.Fatalf("Server has %d connections from the client, want 1", len(server.conns))
+	}
+
+	// Get the server transport for the connecton to the client.
+	var serverTransport *http2Server
+	server.mu.Lock()
+	for k := range server.conns {
+		serverTransport = k.(*http2Server)
+	}
```

Comment:
Do we not need to take `server.mu` before reading `server.conns` (the `len` check on 2155)?

Why a `for` loop instead of simply `serverTransport := server.conns[0].(*http2Server)` (which I'd feel better about with `server.mu` held for the duration of checking `len(server.conns)` and assigning `serverTransport`)?

Reply:
`server.conns ` is of type `map[ServerTransport]net.Conn` so we cannot access a key by an index.

Reply:
`server.mu` moved before reading `server.conns` (the `len` check on 2155)

Reply:
Ah, of course...my mistake.